File: Top-Interview-Questions-Leetcode/Remove-Duplicates-from-Sorted-Array.cpp
#include <iostream>
#include <vector>

class Solution {
        
public:
    int removeDuplicates(std::vector<int>& nums) {
        int cnt = 1, n = nums.size();
        int curIdx = 1, curVal = nums[0];
        
        for (int i = 1; i < n; ++i) {
            if (nums[i] == curVal) {
                nums[i] = -1;
            }
            else {
                curVal = nums[i];
                nums[curIdx++] = nums[i];
                if (curIdx < i) nums[i] = -1;
                cnt++;
                
                while (curIdx < i and nums[curIdx] != -1) curIdx++;
            }
        }
        
        return cnt;
    }
};

// Example usage
int main() {
    Solution solution;
    std::vector<int> nums = {1, 1, 2, 3, 3, 4};
    int newLength = solution.removeDuplicates(nums);
    
    std::cout << "New length: " << newLength << std::endl;
    std::cout << "Modified array: ";
    for (int i = 0; i < newLength; ++i) {
        std::cout << nums[i] << " ";
    }
    std::cout << std::endl;
    
    return 0;
}

File: Top-Interview-Questions-Leetcode/Easy-Collections/Array/Remove-Duplicates-from-Sorted-Array.py
from typing import List

class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        l, sz = 1, len(nums)
        cnt, cur = 1, nums[0]
        
        for h in range(1, sz):
            if nums[h] != cur:
                nums[l] = nums[h]
                l += 1
                cur = nums[h]
                cnt += 1
        
        return cnt

# Example usage
def main():
    nums = [1, 1, 1, 1, 1, 2, 2, 2, 3, 3, 3, 3, 4, 5, 5, 5, 5, 6, 7, 8, 8, 8, 8, 8, 8, 9, 9, 9, 10, 10]
    solution = Solution()
    length = solution.removeDuplicates(nums)
    print(f"Length of array after removing duplicates: {length}")
    print(f"Modified array: {nums[:length]}")

if __name__ == "__main__":
    main()